crates/node-maintainer/src/maintainer.rs

```diff
@@ -173,14 +178,40 @@ impl NodeMaintainer {
         let (package_sink, package_stream) = futures::channel::mpsc::unbounded();
         let mut q = VecDeque::new();
         q.push_back(self.graph.root);
+
+        // Number of dependencies queued for processing in `package_stream`
         let mut in_flight = 0;
 
+        // Since we queue dependencies for multiple packages at once - it is
+        // not unlikely that some of them would be duplicated by currently
+        // fetched dependencies. Thus we maintain a mapping from "name@spec" to
+        // a vector of `NodeDependency`s. When we will fetch the package - we
+        // will apply it to all dependencies that need it.
+        let fetches: HashMap<String, Vec<NodeDependency>> = HashMap::new();
+        let fetches = Arc::new(Mutex::new(fetches));
+
         let mut package_stream = package_stream
             .map(|dep: NodeDependency| {
-                self.nassun
-                    .resolve(format!("{}@{}", dep.name, dep.spec))
-                    .map_ok(move |p| (p, dep))
+                let spec = format!("{}@{}", dep.name, dep.spec);
+                let maybe_spec = if let Some(mut fetches) = fetches.try_lock() {
```

Comment:
Note that I'd use `lock().await` here, but rustc goes bonkers if I do that. Something about MutexGuard not implemented Unpin (which I think it does implement). However, the whole thing happens on a single-thread so the mutex is here only to pacify the borrow-checker anyway.

Reply:
Have you tried to use `async_std::sync::Mutex`? I think that might be the issue.

Reply:
@zkat this is what it was when I got that error. Strange, right?

Reply:
https://gist.github.com/indutny/318abdb226bc951d1ffde25623d42d40

Reply:
That's for:
```rust
                fetches.lock().map(|mut fetches| {
                    if let Some(list) = fetches.get_mut(&spec) {
                        // Package fetch is already in-flight, add dependency
                        // to the existing list.
                        list.push(dep);
                        None
                    } else {
                        // Fetch package since we are the first one to get here.
                        fetches.insert(spec.clone(), vec![dep]);
                        Some(spec)
                    }
                })
```

Reply:
oh. Yeah, that's weird. I have no idea. Pinning is like, beyond me.

Reply:
Same, it didn't work anyway. I can make a bit of progress with `Box::pin`, but it then fails at "cannot prove".

Reply:
"cannot prove" is hell. Are you gonna keep trying, or do you want me to merge this? :)

Reply:
Let's merge this and get done with it :)